pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,215 @@
+    retain_grads(z_next)
+    retain_grads(r_next)
+    grads = grad_potential(z_next)
+
+    for _ in range(num_steps):
+        # detach graph nodes for next iteration
+        detach_nodes(z_next)
+        detach_nodes(r_next)
+        for site_name in z_next:
+            # r(n+1/2)
+            r_next[site_name] = r_next[site_name] + 0.5 * step_size * (-grads[site_name])
+            # z(n+1)
+            z_next[site_name] = z_next[site_name] + step_size * r_next[site_name]
+        # retain gradients for intermediate nodes in backward step
+        retain_grads(z_next)
+        retain_grads(r_next)
+        grads = grad_potential(z_next)
+        for site_name in r_next:
+            # r(n+1)
+            r_next[site_name] = r_next[site_name] + 0.5 * step_size * (-grads[site_name])
+    return z_next, r_next
+
+
+def retain_grads(z):
+    for key in z:
+        # XXX: can be removed with PyTorch 0.3
+        if z[key].is_leaf and not z[key].requires_grad:
+            z[key].requires_grad = True
+        z[key].retain_grad()
+
+
+def detach_nodes(z):
+    for key in z:
+        z[key] = Variable(z[key].data, requires_grad=True)
+
+
+class HMC(TracePosterior):
+    def __init__(self,
+                 model,
+                 step_size=0.5,
+                 num_steps=3,
+                 warmup_steps=20,
+                 num_samples=1000):
+        self.model = model
+        self.step_size = step_size
+        self.num_steps = num_steps
+        self.num_samples = num_samples
+        self.warmup_steps = warmup_steps
+        # simulation run attributes - will be set in self._setup
+        # at start of run
+        self.cached_param_grads = {}
+        self.args = None
+        self.kwargs = None
+        self.accept_cnt = None
+        self.prototype_trace = None
+        super(HMC, self).__init__()
+
+    def log_prob(self, z):
+        """
+        Return log pdf of the model with sample sites replayed from z_trace
+        """
+        z_trace = self.prototype_trace.copy()
+        for name, value in z.items():
+            z_trace.nodes[name]['value'] = value
+        model_trace = poutine.trace(poutine.replay(self.model, trace=z_trace)) \
+            .get_trace(*self.args, **self.kwargs)
+        return model_trace.log_pdf()
+
+    def grad_potential(self, z):
+        log_joint_prob = self.log_prob(z)
+        log_joint_prob.backward()
+        grad_potential = {}
+        for name, value in z.items():
+            grad_potential[name] = -value.grad.clone().detach()
+            grad_potential[name].volatile = False
+        return grad_potential
+
+    def _setup(self, *args, **kwargs):
+        self.accept_cnt = 0
+        self.args = args
+        self.kwargs = kwargs
+        # set the trace prototype to inter-convert between trace object
+        # and dict object used by the integrator
+        self.prototype_trace = poutine.trace(self.model).get_trace(*args, **kwargs)
```

Comment:
We should have some asserts here that check that there are no non-reparametrizable variables and that `self.prototype_trace.log_pdf() > 0`

Reply:
Good point, will add this.

Reply:
Sorry, I mean `self.prototype_trace.log_pdf() > -infinity`